### PR TITLE
Order model

### PR DESCRIPTION
**Pull request by RomanYuzviak**

### CODE REVIEW

**Review by KDanil2000 — APPROVED** [review]
LGTM!

File: src/main/java/com/example/onlinebookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+public class OrderController {
+    private final OrderService orderService;
+    private final UserRepository userRepository;
+
+    @GetMapping
+    @ResponseStatus(HttpStatus.OK)
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "get all orders for the user")
+    public List<OrderDto> getAll(Pageable pageable, Principal principal) {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/controller/OrderController.java]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+@RequiredArgsConstructor
+public class OrderController {
+    private final OrderService orderService;
+    private final UserRepository userRepository;
+
+    @GetMapping
+    @ResponseStatus(HttpStatus.OK)
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "get all orders for the user")
+    public List<OrderDto> getAll(Pageable pageable, Principal principal) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/controller/OrderController.java:46]: You can use Authentication authentication like method parameter, then getting user by
User user = (User) authentication.getPrincipal();
and use user.getId();

it's avoid unnecessary query to database in method private Long getUserId(Principal principal) {

File: src/main/java/com/example/onlinebookstore/controller/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    public OrderDto patchOrder(@PathVariable @Positive Long id,
+                               @RequestBody @Valid OrderUpdateDto updateDto) {
+        return orderService.update(updateDto, id);
+    }
+
+    @GetMapping("{orderId}/items")
+    @ResponseStatus(HttpStatus.OK)
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "get all items for the order")
+    public List<OrderItemDto> getAllItems(@PathVariable @Positive Long orderId, Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/controller/OrderController.java]:
```suggestion
    public List<OrderItemDto> getAllItems(@PathVariable @Positive Long orderId, @ParameterObject @PageableDefault Pageable pageable) {
```

File: src/main/java/com/example/onlinebookstore/dto/order/OrderUpdateDto.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package com.example.onlinebookstore.dto.order;
+
+import com.example.onlinebookstore.model.Order;
+import jakarta.validation.constraints.NotBlank;
+
+public record OrderUpdateDto(
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/order/OrderUpdateDto.java]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/dto/order/OrderUpdateDto.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package com.example.onlinebookstore.dto.order;
+
+import com.example.onlinebookstore.model.Order;
+import jakarta.validation.constraints.NotBlank;
+
+public record OrderUpdateDto(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/order/OrderUpdateDto.java]: It's request or response?

File: src/main/java/com/example/onlinebookstore/dto/order/OrderRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.example.onlinebookstore.dto.order;
+
+import jakarta.validation.constraints.NotBlank;
+
+public record OrderRequestDto(
+        @NotBlank
+        String shippingAddress
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/order/OrderRequestDto.java:7]: Do you need Size annotation?

File: src/main/java/com/example/onlinebookstore/mapper/OrderMapper.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import com.example.onlinebookstore.dto.order.OrderUpdateDto;
+import com.example.onlinebookstore.model.Order;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+
+@Mapper(config = MapperConfig.class,
+        uses = OrderItemMapper.class)
+public interface OrderMapper {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/mapper/OrderMapper.java]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/model/Order.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+@Table(name = "orders")
+@Data
+@SQLDelete(sql = "UPDATE orders SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Order.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
```

File: src/main/java/com/example/onlinebookstore/model/Order.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import java.time.LocalDateTime;
+import java.util.List;
+import lombok.Data;
+import lombok.ToString;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Table(name = "orders")
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Order.java]:
```suggestion
@Getter
@Setter
```

File: src/main/java/com/example/onlinebookstore/model/Order.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Column(nullable = false)
+    private String shippingAddress;
+
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    @OneToMany(cascade = CascadeType.ALL, mappedBy = "order",
+            fetch = FetchType.LAZY)
+    @ToString.Exclude
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Order.java]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/model/OrderItem.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+@SQLDelete(sql = "UPDATE order_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    @ToString.Exclude
+    private Order order;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "book_id", nullable = false)
```

File: src/main/java/com/example/onlinebookstore/model/OrderItem.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+@Entity
+@Table(name = "order_items")
+@Data
+@SQLDelete(sql = "UPDATE order_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "order_id", nullable = false)
```

File: src/main/java/com/example/onlinebookstore/model/OrderItem.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+import jakarta.persistence.Table;
+import java.math.BigDecimal;
+import lombok.Data;
+import lombok.ToString;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Table(name = "order_items")
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/OrderItem.java]:
```suggestion
@Getter
@Setter
```

File: src/main/java/com/example/onlinebookstore/model/OrderItem.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+@SQLDelete(sql = "UPDATE order_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    @ToString.Exclude
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/OrderItem.java]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/repository/OrderRepository.java [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package com.example.onlinebookstore.repository;
+
+import com.example.onlinebookstore.model.Order;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface OrderRepository extends JpaRepository<Order, Long> {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/repository/OrderRepository.java:8]: `@Repository`

File: src/main/java/com/example/onlinebookstore/dto/order/OrderUpdateDto.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+        @NotBlank
+        String shippingAddress,
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/order/OrderUpdateDto.java]: you don't need shippingAddress in this request. The requirements are said about update status

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    @Override
+    public List<OrderDto> findAll(Pageable pageable, Long userId) {
+        return orderRepository.findAll(pageable).stream()
+                .map(orderMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    @Transactional
+    public OrderDto update(OrderRequestDto requestDto, Long userId) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]: Why you have two update methods? What is the logic about?

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+        orderItems.forEach(oi -> oi.setOrder(newOrder));
+        newOrder.setOrderItems(orderItems);
+        return orderMapper.toDto(
+                orderRepository.save(newOrder)
+        );
+    }
+
+    @Override
+    public OrderDto update(OrderUpdateDto updateDto, Long orderId) {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+        Order currentOrder = orderRepository.findById(orderId)
+                .orElseThrow(() -> new EntityNotFoundException("Order is not found"));
+        orderMapper.updateOrder(updateDto, currentOrder);
+
+        return orderMapper.toDto(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]:
```suggestion
        return orderMapper.toDto(orderRepository.save(currentOrder));
```

File: src/main/java/com/example/onlinebookstore/mapper/OrderItemMapper.java [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+import com.example.onlinebookstore.config.MapperConfig;
+import com.example.onlinebookstore.dto.order.OrderItemDto;
+import com.example.onlinebookstore.model.CartItem;
+import com.example.onlinebookstore.model.OrderItem;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
+public interface OrderItemMapper {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/mapper/OrderItemMapper.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/com/example/onlinebookstore/dto/order/OrderUpdateDto.java (hunk 16 above)
Comment (RomanYuzviak) [reply src/main/java/com/example/onlinebookstore/dto/order/OrderUpdateDto.java]: I thought it`ll be good to allow admin to change an address

File: src/main/java/com/example/onlinebookstore/model/Order.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE orders SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Order.java:35]: `@JoinColumn(name = "user_id", nullable = false)`

File: src/main/java/com/example/onlinebookstore/model/OrderItem.java [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+@Table(name = "order_items")
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE order_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/OrderItem.java:28]: `@JoinColumn(name = "order_id", nullable = false)`

File: src/main/java/com/example/onlinebookstore/model/OrderItem.java [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+    private Order order;
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/OrderItem.java:31]: `@JoinColumn(name = "book_id", nullable = false)`

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+                orderRepository.save(newOrder)
+        );
+    }
+
+    public OrderDto update(OrderUpdateRequestDto updateDto, Long orderId) {
+        Order currentOrder = orderRepository.findById(orderId)
+                .orElseThrow(() -> new EntityNotFoundException("Order is not found"));
+        orderMapper.updateOrder(updateDto, currentOrder);
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]: remove empty line

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+                }
+                )
+                .toList();
+        newOrder.setTotal(orderItems.stream()
+                .map(oi -> oi.getPrice().multiply(BigDecimal.valueOf(oi.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add)
+        );
+        orderItems.forEach(oi -> oi.setOrder(newOrder));
+        newOrder.setOrderItems(orderItems);
+        return orderMapper.toDto(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java]:
```suggestion
        return orderMapper.toDto(orderRepository.save(newOrder));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

**Review by liliia-ponomarenko — APPROVED** [review]
Great! But fix the conflict before merging;)

**Pull request closed without merging**